### ISSUE TITLE
Internal compiler error in FrontendAction::Execute

### ISSUE THREAD

**Issue by bbannier** [llvm/llvm-project#32929]
|  |  |
| --- | --- |
| Bugzilla Link | [33582](https://llvm.org/bz33582) |
| Version | trunk |
| OS | MacOS X |
| Attachments | [Source file](https://user-images.githubusercontent.com/95090706/143755881-48919683-0ad1-4261-8abc-410c7aca230e.gz), [Run file](https://user-images.githubusercontent.com/95090706/143755883-ad078e6f-aac7-44f7-acfd-1ae714b2bf5c.gz), [Crash report](https://user-images.githubusercontent.com/95090706/143755884-4eb6e839-c955-4593-98ca-cc9687409dd5.gz) |

## Extended Description 
I hit a compiler segmentation fault on macOS Sierra with the following version.

% clang --version
clang version 5.0.0 (http://llvm.org/git/clang 283ac36731893129f189c580dbe2b4b1f6e87241) (http://llvm.org/git/llvm 472c4c70536a09587d06fce0b18ef4ec07600434)
Target: x86_64-apple-darwin16.6.0
Thread model: posix
InstalledDir: /Users/bbannier/src/llvm/P/bin

  Registered Targets:
    x86    - 32-bit X86: Pentium-Pro and above
    x86-64 - 64-bit X86: EM64T and AMD64

libcxx: trunk@306083

I would instead have expected processing of the file, and either a diagnostic message or an executable output file.

**Comment by Endilll** [comment]
Appears to be fixed in Clang 9: https://godbolt.org/z/nv71bTfza
Reduced by C-Reduce:
```cpp
template < bool , class _Tp >
using enable_if_t = _Tp;

template < int  >
struct integral_constant 

template < class _Tp >
_Tp __declval;

template < class _Tp >
struct __invokable_r  {
  using _Result = decltype(declval< _Fp >());
}

template < class , class >
struct is_callable

template < class _Fn, class _Ret >
struct is_callable< _Fn(), _Ret > : integral_constant< __invokable_r<  _Fn>::value > {}

template < class _Fn, class _Ret = void >
constexpr bool is_callable_v = is_callable< _Fn, _Ret >::value

template < typename Fn >
auto value(Fn ) -> enable_if_t< is_callable_v< Fn() >, Fn >

template < typename Fn >
struct fn_t : Fn {
  fn_t(Fn ) {
    auto fn = [](auto ...) {
      return Fn::operator();
    };
    value(fn_t< decltype(fn) >{fn});
  }
}

main0 {
  fn_t{[] {}}
```
